bundles/org.openhab.binding.hive/src/main/java/org/openhab/binding/hive/internal/handler/HiveBridgeHandler.java

```diff
@@ -0,0 +1,394 @@
+                    if (node.attributes != null && node.attributes.batteryLevel != null) {
+                        // If this a thermostat, and it has a current temperature, add it to discovery results
+                        systemInfo = node;
+                    }
+                }
+                if (thermostat != null) {
+                    if (systemInfo != null) {
+                        thermostat.linkedNode = systemInfo.id;
+                        thermostat.firmwareVersion = systemInfo.attributes.softwareVersion.displayValue;
+                        thermostat.model = systemInfo.attributes.model.displayValue;
+                        thermostat.macAddress = systemInfo.attributes.macAddress.displayValue;
+                    }
+                    if (hiveDiscoveryService != null) {
+                        hiveDiscoveryService.addDevice(thermostat);
+                    }
+                }
+            }
+        }
+    }
+
+    public HiveAttributes getThermostatReading(Thing thing) {
+        HiveAttributes reading = new HiveAttributes();
+
+        if (online) {
+            ContentResponse response;
+            int statusCode = 0;
+            String responseString = "";
+            HiveNodes o = null;
+
+            // Get thermostat reading
+            try {
+                response = client.newRequest("https://api-prod.bgchprod.info:443/omnia/nodes/" + thing.getUID().getId())
+                        .method(HttpMethod.GET).header("Accept", "application/vnd.alertme.zoo-6.1+json")
+                        .header("Content-Type", "application/vnd.alertme.zoo-6.1+json")
+                        .header("X-Omnia-Client", "Openhab 2").header("X-Omnia-Access-Token", token)
+                        .timeout(DISCOVER_TIMEOUT_SECONDS, TimeUnit.SECONDS).send();
+
+                statusCode = response.getStatus();
+
+                if (statusCode == HttpStatus.UNAUTHORIZED_401) {
+                    // Token expired, get a new one and try again
+                    getToken();
+                    statusCode = response.getStatus();
+                }
+
+                if (statusCode != HttpStatus.OK_200) {
+                    // If it failed, log the error
+                    String statusLine = response.getStatus() + " " + response.getReason();
+                    logger.warn("Error while reading from Hive API: {}", statusLine);
+                    return reading;
+                }
+
+                responseString = response.getContentAsString();
+                o = gson.fromJson(responseString, HiveNodes.class);
+                reading = o.nodes.get(0).attributes;
+                reading.isValid = true;
+            } catch (InterruptedException | TimeoutException | ExecutionException e) {
+                logger.warn("Failed to get thermostat reading: {}", e.getMessage());
+                return reading;
+            }
+            try {
+                // Get battery level
+                response = client
```

Comment:
Maybe you should split this get with a dynamic part of the url of as a separate method including the HTTP status code handling and JSON mapping, by passing a generic:

it looks very similar to `callClient`:

private <T> T getFromClient(String url,  Class<T> class)